### PR TITLE
[WIP] Service Fabric support

### PR DESCRIPTION
**Pull request by ReubenBond**
This is an implementation of deeper Service Fabric support (#1059).
It's not ready for use, so please don't use it. I'm pushing an early revision so that people can see what this might look like. It will change significantly before it's merged and may be split into multiple PRs.

* `OrleansCommunicationListener` + a helper class, `OrleansServiceListener`: Allows users to host a silo in a Stateful or Stateless Reliable Service (Service Fabric parlance). Silos can co-exist with Web APIs, WCF services, etc. The silo will publish an endpoint called "Orleans" which is used when locating the service.
* `OrleansServiceFabricExtensions`: Helper class which adds an `.AddServiceFabricSupport()` extension method to `IServiceCollection` so that the relevant services can be made available via dependency injection.
* `ReliableDictionaryStateProvider`: an `IStorageProvider` which replicates state from the local primary to remote secondary replicas.
* (internal) `ServiceFabricNamingServiceGatewayProvider`: `IMembershipTable` & `IGatewayListProvider` implementation which uses SF's internal strong-consistency Naming Service as the store, eliminating the need for Azure Tables. SF's Naming Service is similar to Zookeeper.
* (internal) `ServiceFabricPartitionPlacementDirector`: Placement director which maps grains to SF partitions (and hence silos) based on ID. This allows state state
* (internal) `PartitionResolver`: Helper class which maps SF partitions to Orleans silo endpoints. Used by the placement provider.
* (internal) `OrleansFabricSilo`: Internal silo management class which hosts the Orleans silo instance and manages its lifecycle in accordance with SF's lifecycle notifications (via `OrleansCommunicationListener`.

I made changes to DI which will not be merged - I am using a delegate to configure DI, but it looks as though we will mirror ASP.NET's system instead. The projects in the Samples folder will be placed in their own solutions (and will be moved around a bit).

Before this is ready to be merged, it is crucial to ensure that the placement director works as advertised and that grains can never migrate to another Service Fabric partition. SF partition counts/names are fixed once deployed, but a single executable can host multiple partitions and SF can migrate partitions between nodes during failover and for load balancing. With Orleans currently having many static classes and being unable to host multiple silos in the same AppDomain, this PR cannot guarantee that a grain is actually on the primary for its partition.  This also lacks tests.

### CODE REVIEW

**Comment by ReubenBond** [comment]
If I make a `feature-service-fabric` branch, then will automated tests still run? If so, I will do that so that we can develop this there.

**Comment by sergeybykov** [comment]
I don't believe CI will run for a branch other than `master`. I don't know if that's something that's easy to configure somewhere in Jenkins' config.

File: src/OrleansServiceFabricUtils/ServiceFabricNamingServiceGatewayProvider.cs [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+
+        private async Task<MembershipTableData> ReadEntries(SiloAddress siloAddress = null)
+        {
+            ReadResults result;
+            do
+            {
+                // Continue attempting to read from the table until a consistent read is made.
+                result = await this.TryReadEntries(siloAddress);
+            }
+            while (!result.IsConsistent);
```
Comment (gabikliot) [inline src/OrleansServiceFabricUtils/ServiceFabricNamingServiceGatewayProvider.cs]: some read may be inconsistent? really? how? I thought the bread and butter of SF and its primary pride was strong consistency.

**Comment by gabikliot** [comment]
>this PR cannot guarantee that a grain is actually on the primary for its partition.

Why?

File: src/OrleansServiceFabricUtils/ServiceFabricNamingServiceGatewayProvider.cs [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+namespace Microsoft.Orleans.ServiceFabric
+{
+    using System.Fabric;
+    using System.Globalization;
+
+    using Newtonsoft.Json;
+    using Newtonsoft.Json.Converters;
+    using Newtonsoft.Json.Linq;
+
+    public class ServiceFabricNamingServiceGatewayProvider : IMembershipTable, IGatewayListProvider
```
Comment (gabikliot) [inline src/OrleansServiceFabricUtils/ServiceFabricNamingServiceGatewayProvider.cs]: The name confused me at first. Its both Table and GW Provider, not just GW provider.

File: src/OrleansServiceFabricUtils/ReliableDictionaryStateProvider.cs [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+            return Task.FromResult(0);
+        }
+
+        public Logger Log { get; private set; }
+
+        public async Task ReadStateAsync(string grainType, GrainReference grainReference, IGrainState grainState)
+        {
+            var store = await this.GetStore();
+            ConditionalValue<GrainState> result;
+            using (var tx = this.stateManager.CreateTransaction())
```
Comment (gabikliot) [inline src/OrleansServiceFabricUtils/ReliableDictionaryStateProvider.cs]: why do you need a transaction around it?

File: src/OrleansServiceFabricUtils/ServiceFabricNamingServiceGatewayProvider.cs (hunk 1 above)
Comment (ReubenBond) [inline src/OrleansServiceFabricUtils/ServiceFabricNamingServiceGatewayProvider.cs]: It looks like the Naming Service in particular allows for torn reads. It notifies you when they occur, at least. We will retry until we read a consistent view of the entire table.

Comment (gabikliot) [inline src/OrleansServiceFabricUtils/ServiceFabricNamingServiceGatewayProvider.cs]: I am sure you know that already, but you could have also integrated differently: instead of using NS as a key value store for MBR table, use NS as a full blown membership provider: listen to its notification about silo liveness, read all silos from it, use the SF internal membership basically and fully replace Orleans' MBR. Deeper integration sort of.
That's what I did 3-4 years ago.

**Comment by ReubenBond** [comment]
@gabikliot In its current iteration if there are multiple partition on one node, then it will try to start multiple silos in the same AppDomain, which I believe will result in a failure. My original comment was wrong - I think it will fail to initialize rather than initialize but behave incorrectly.

My original (flawed) thinking was this:
* Multiple SF partitions on one node (overprovisioning is common in SF and regardless, failovers can result in multiple primaries for different partitions on one node)
* Single process per node, regardless of how many partitions. This is SF behavior, maybe it can be changed.
* This PR maps GrainId -> PartitionId -> SiloAddress
* SF can migrate a partition at runtime (eg, failover, load balancing).
* Given a silo could be the primary for multiple partitions, a mapping from GrainId -> PartitionId -> SiloAddress could still map to an active primary (for the wrong partition) even though the intended partition has changed its primary.

In short: need to ensure that a SiloAddress is only ever mapped to one partition. Also need to ensure that if a silo's generation doesn't match the generation specified by the sender, then it rejects the message. Maybe those conditions are already true.

File: src/OrleansServiceFabricUtils/ServiceFabricNamingServiceGatewayProvider.cs (hunk 2 above)
Comment (ReubenBond) [inline src/OrleansServiceFabricUtils/ServiceFabricNamingServiceGatewayProvider.cs]: Maybe I should rename this to `ServiceFabricNamingServiceClusterMembershipProvider`. The Zookeeper provider also serves the same dual purpose of being a membership table & gateway list provider.

File: src/OrleansServiceFabricUtils/ReliableDictionaryStateProvider.cs (hunk 3 above)
Comment (ReubenBond) [inline src/OrleansServiceFabricUtils/ReliableDictionaryStateProvider.cs]: `IReliableDictionary`'s methods require a transaction.

File: src/OrleansServiceFabricUtils/ServiceFabricNamingServiceGatewayProvider.cs (hunk 1 above)
Comment (ReubenBond) [inline src/OrleansServiceFabricUtils/ServiceFabricNamingServiceGatewayProvider.cs]: I may add a version which does what you did, but for now I wanted to adhere to Orleans' membership protocol as tightly as possible.

**Comment by gabikliot** [comment]
>Given a silo could be the primary for multiple partitions, a mapping from GrainId -> PartitionId -> SiloAddress could still map to an active primary (for the wrong partition) even though the intended partition has changed its primary.

**Comment by gabikliot** [comment]
> Given a silo could be the primary for multiple partitions, a mapping from GrainId -> PartitionId -> SiloAddress could still map to an active primary (for the wrong partition) even though the intended partition has changed its primary.

I am  not sure I understand that.

>need to ensure that a SiloAddress is only ever mapped to one partition

Why?

I thought we map grains ids to partitions, and SF maps partitions to silo (service instances) based on latest membership it mainatins and SF guarantees that any partition is mapped to at most single service instance at a given time. That is the strong consistency property that I thought it provides.
What am I missing?

**Comment by gabikliot** [comment]
Found it: https://github.com/dotnet/orleans/blob/master/src/OrleansRuntime/Messaging/IncomingMessageAcceptor.cs#L439
We simply drop msg if they are for my silo ip:port but different gen number. super important as part of ensuring "enforcing perfect failure detection".

`SA.Matches():` Gen id is zero for msgs from the client, since client, by design, does not stick to a particular silo but rather to any GW.  That specific design point about client-gw is not actually too fundamental, can be changed, its more about separation of concerns.

**Comment by jdom** [comment]
I can very easily configure Jenkins to build a branch other than `master` if that's what we want here

**Comment by ReubenBond** [comment]
@jdom that would be great. I'd like to wait for the new DI config method before merging this into its own branch, so I'll leave it for now

**Comment by ReubenBond** [comment]
I plan on separating this out into smaller PRs so that we can merge the functionality present in OrleansContrib/OrleansFabricSilo sooner as we incrementally add deeper Service Fabric support.

**Comment by Kwal** [comment]
@ReubenBond do you have any thoughts as to which piece you're going to create a PR for first?

My vote would be for the `IMembershipTable` and `IGatewayListProvider` implementation(s) - I think that'd be a huge win for running in SF and it seems like that piece is fairly straightforward (based on your existing code).

**Comment by ReubenBond** [comment]
@Kwal that's the plan :) 
#2120 first, then membership/gateway, then placement, storage, etc
